Fixed DSR loopback connectivity for Windows

zawachte opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bugfix


**What this PR does / why we need it**:
Fixed DSR loopback connectivity for Windows

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE

```

liggitt committed:
Fix race in quota sync test

Commit:
kubeadm: add final fallback to constants.CurrentKubernetesVersion

It may happen that both the git version and the remote version
are broken/inaccessible. In this case the broken remote version
would be used.

To overcome this situation fall back to the constant CurrentKubernetesVersion.

The alternative could be os.Exit(1).

Also this change fixes Bazel-based unit tests in air-gapped environment.

Commit:
resourcequota: use dynamic informer

The resource quota controller should use a dynamic informer so it
can create informer for custom resources.

giuseppe committed:
kubelet: read immediately after creating the watcher

if some events happen between the Read and while we set the watcher,
we might miss them.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

giuseppe committed:
kubelet: handle recreated log files

if the runtime is configured to rotate the log file, we might end up
watching the old fd where there are no more writes.

When a fsnotify event other than Write is received, reopen the log
file and recreate the watcher.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

giuseppe committed:
kubelet: force checking the log file every second

it seems fsnotify can miss some read events, blocking the kubelet to
receive more data from the log file.

If we end up waiting for events with fsnotify, force a read from the
log file every second so that are sure to not miss new data for longer
than that.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

giuseppe committed:
logs: consume all file until EOF on exited container

If the container is not found, do not stop reading the log file
immediately but wait until we reach again EOF.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

vllry committed:
Amend TODO comment about ClientConnectionConfiguration's types (#73375)

vllry committed:
Updated kube-proxy error strings to not use capitals/punctuation

vllry committed:
Remove redundant import names in kube-proxy

vllry committed:
Renamed variables colliding with package names, and added TODOs on uncaught-errors

WanLinghao committed:
Remove dead code about failure-domains option in kube-scheduler

odino committed:
Attaching owner references to jobs create via "--from=cronjob/xyz"

smarterclayton committed:
Report a watch error instead of eating it when we can't decode

Clients are required to handle watch events of type ERROR, so instead
of eating the decoding error we should pass it on to the client. Use
NewGenericServerError with isUnexpectedResponse to indicate that we
didn't get the bytes from the server we were expecting. For watch, the
415 error code is roughly correct and we will return an error to the
client that makes debugging a failure in either server watch or client
machinery much easier.

We do not alter the behavior when it appears the response is an EOF
or other disconnection.

Commit:
delete unused code in proxy_server.go

WanLinghao committed:
Migrate the controller to use TokenRequest and rotate token periodically

paulsubrata55 committed:
Fix for issue #73300. kube-proxy with IPVS and sctp traffic

wojtek-t committed:
Reduce critical sections in watchache

Commit:
Rename RunPullImagesCheck to PullControlPlaneImages

Commit:
Implement CSI migration logic for block volume

ialidzhikov committed:
Update fluentd to 1.4.0

Signed-off-by: ialidzhikov <i.alidjikov@gmail.com>

pmcalpine committed:
Fix lint warnings in pkg/controller/bootstrap

pmcalpine committed:
Rename: cronjob_controller[_test].go -> controller[_test].go

This change renames these source files to reduce stuttering when reading
the path.

pmcalpine committed:
Fix lint warnings in pkg/controller/cronjob.

mkimuram committed:
Add e2e tests for multiAttach

WanLinghao committed:
Remove 404 link in kubectl help information

rramkumar1 committed:
Remove tests exercising the backend protocol feature of Ingress-GCE

ialidzhikov committed:
Update golang to 1.12.0

Signed-off-by: ialidzhikov <i.alidjikov@gmail.com>

Commit:
fix a mistake

deads2k committed:
make kubectl scale work without a GET if a precodition isn't requested

jb-perez committed:
Bump debian-base images version to v1.0.0

Change-Id: I6b924f0102f3c8490c6f71efb471ebe08827d712

vllry committed:
Merge branch 'master' into kube-proxy-cleanup

liggitt committed:
Clean up test output

liggitt committed:
Prevent quota controller using unsynced listers

liggitt committed:
Improve quota sync log messages

liggitt committed:
Avoid deadlock in resource quota resync

liggitt committed:
Update quota status with limits even when calculating errors

liggitt committed:
Only reject quota admission if status is missing relevant usage

liggitt committed:
Generate networkpolicies extensions/v1beta1 client

liggitt committed:
Generated files

vllry committed:
Fix a bad merge leftover in kube-proxy

serathius committed:
Update gcp images with security patches

[stackdriver addon] Bump prometheus-to-sd to v0.5.0 to pick up security fixes.
[fluentd-gcp addon] Bump fluentd-gcp-scaler to v0.5.1 to pick up security fixes.
[fluentd-gcp addon] Bump event-exporter to v0.2.4 to pick up security fixes.
[fluentd-gcp addon] Bump prometheus-to-sd to v0.5.0 to pick up security fixes.
[metatada-proxy addon] Bump prometheus-to-sd v0.5.0 to pick up security fixes.

sttts committed:
apiextensions: merge OpenAPI specs once after cache sync

shinytang6 committed:
replace time.Now().Sub with time.Since

xichengliudui committed:
Add xichengliudui to test/OWNERS

mucahitkurt committed:
add expectedCSINode and expectedNode for test cases to compare CSINode and Node object instead of field based comparisons

Signed-off-by: Mucahit Kurt <mucahitkurt@gmail.com>

gsadhani committed:
Skip smart label in kubectl describe if field has special chars

Attempts to create smart label for fields containing special chanracter ends
up looking very odd. This change skips creating smart labels for fields
containing special characters other than '-'.

hex108 committed:
Pick pods for preemption based on StartTime of pods when priorities are equal

Commit:
Kubectl: support inline csi volume

dougm committed:
godeps: update vmware/govmomi to v0.20 release

dougm committed:
vSphere: add token auth support for tags client

SAML auth support for the vCenter rest API endpoint came to govmomi
a bit after Zone support came to vSphere Cloud Provider.

Fixes #75511

dougm committed:
vsphere: govmomi rest API simulator requires authentication

xichengliudui committed:
Fix shellcheck failures in godep-save.sh and godep-restore.sh

update pull request

update pull request

dims committed:
New staging repository for cri-api

Change-Id: I2160b0b0ec4b9870a2d4452b428e395bbe12afbb

xichengliudui committed:
Define common sentences as constants

update pull request

update pull request

update pull request

update pull request

update pull request

update pull request

tiffanyfay committed:
Remove aws cred provider dep on cloud provider

tiffanyfay committed:
Credential provider Provide takes image (general)

tiffanyfay committed:
Credential provider Provide takes image (clouds)

tiffanyfay committed:
Refactors and fixes bugs in AWS credentialprovider

Adds caching per registry. Fixes caching of invalid ECR tokens.

liggitt committed:
Add preliminary go module instructions to client-go

Commit:
Implementing DSR by adding endpoint policy

Implementing DSR by adding endpoint policy

PR feedback

works

Implementing DSR by adding endpoint policy

Implementing DSR by adding endpoint policy

Implementing DSR by adding endpoint policy

PR feedback

switching branches

Implementing DSR by adding endpoint policy

Implementing DSR by adding endpoint policy

PR feedback

switching branches

Commit:
Implementing DSR by adding endpoint policy

Commit:
Implementing DSR by adding endpoint policy

Commit:
Implementing DSR by adding endpoint policy

smarterclayton committed:
Check CRD protobuf negotiation in integration tests

Add a few more tests to verify CRDs don't regress as client changes
are made.

tiffanyfay committed:
Update aws provider build files

tallclair committed:
Copy node test owners to e2e/common

Commit:
Fix golint failures of e2e/framework/[g-j]*.go

Commit:
Fix golint failures of e2e/framework/[k-l]*.go

In addition, this replaces the combination of GetCPUSummary() and
FormatCPUSummary() with LogCPUSummary() in e2e/node/kubelet_perf.go
because that is completely same.

SataQiu committed:
fix some failures in pkg/apis

ialidzhikov committed:
Fix lint issues

Signed-off-by: ialidzhikov <i.alidjikov@gmail.com>

bart0sh committed:
kubeadm: implement deletion of multiple tokens

Commit:
sample-controller: update README

Commit:
fix eliminate shellcheck failures in hack/local-up-cluster.sh

bart0sh committed:
fix 'make generated_files' build on MacOS

Recent change to hack/lib/golang.sh broke the build on MacOS this way:

$ make clean && make generated_files
+++ [0325 13:38:22] Verifying Prerequisites....
+++ [0325 13:38:23] Removing _output directory
k8s.io/kubernetes/vendor/github.com/spf13/pflag
k8s.io/kubernetes/hack/make-rules/helpers/go2make
+++ [0325 13:38:40] Building go targets for darwin/amd64:
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
can't load package: package k8s.io/kubernetes: no Go files in k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes
!!! [0325 13:38:40] Call tree:
!!! [0325 13:38:40]  1: k8s.io/kubernetes/hack/lib/golang.sh:629 kube::golang::build_some_binaries(...)
!!! [0325 13:38:40]  2: k8s.io/kubernetes/hack/lib/golang.sh:764 kube::golang::build_binaries_for_platform(...)
!!! [0325 13:38:40]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make[1]: *** [_output/bin/deepcopy-gen] Error 1
make: *** [generated_files] Error 2

It was caused by 'binaries' array not being declared with 'local -a'.
It looks like MacOS' old bash version makes an array to contain first
empty element if declared this way.

The fix has been tested on MacOS High Sierra and Linux openSUSE 42.3 (x86_64)

Signed-off-by: Ed Bartosh <eduard.bartosh@intel.com>

fabriziopandini committed:
e2e-kubeadm-refactor

jlucktay committed:
Enact golint recommendations per issue 68026.

Remove the 'abac' package from the golint exclusion list.
Add/edit comments per golint feedback.
Set PolicyList to be exported, as not exporting was breaking one of
golint's rules around exported funcs returning an unexported type.
Fix a broken test

tallclair committed:
Refactor PSP provider

tallclair committed:
Modernize PSP unit test

Commit:
Fix golint failures of e2e/framework/[d-e]*.go

This fixes golint failures on the following files:
- test/e2e/framework/deployment_util.go
- test/e2e/framework/exec_util.go

Cleanup:
- ScaleDeployment() was not used at all, so let's remove it.
- ExecCommandInPod() and ExecCommandInPodWithFullOutput() were called
  in the framework only, so let's make them local.

freehan committed:
make NEG tests more resilient to API failures and add more failure logging

smarterclayton committed:
Reduce allocations in metav1.Time JSON serialization

Time has a limited set of allowed characters and so can bypass part of
the expensive checking done for valid JSON characters.

```
BenchmarkGet-12          	   10000	    111846 ns/op	   17273 B/op	     130 allocs/op
BenchmarkWatchHTTP-12    	   50000	     37830 ns/op	    1886 B/op	      29 allocs/op

BenchmarkGet-12          	   20000	    115917 ns/op	   17344 B/op	     130 allocs/op
BenchmarkWatchHTTP-12    	   50000	     36741 ns/op	    1775 B/op	      24 allocs/op
```

Can improve CRD watch performance by 5-10% for small objects.

zawachte-msft commented:
@madhanrm @ksubrmnn

ksubrmnn commented on pkg/proxy/winkernel/hnsV1.go:
```diff
@@ -35,6 +35,7 @@ type HostNetworkService interface {
 	deleteEndpoint(hnsID string) error
 	getLoadBalancer(endpoints []endpointsInfo, isILB bool, isDSR bool, sourceVip string, vip string, protocol uint16, internalPort uint16, externalPort uint16) (*loadBalancerInfo, error)
 	deleteLoadBalancer(hnsID string) error
+	updateEndpointPolicy(hnsID string, policy json.RawMessage) error
```
Can you change hnsID to endpointID so that it is more clear?

ksubrmnn commented on pkg/proxy/winkernel/hnsV1.go:
```diff
@@ -223,3 +224,7 @@ func (hns hnsV1) deleteLoadBalancer(hnsID string) error {
 }
+
+func (hns hnsV1) updateEndpointPolicy(hnsID string, policy json.RawMessage) error {
+	return fmt.Errorf("DSR is not supported in V1. Using non DSR instead")
```
Change this to updateEndpointPolicy is not implemented for V1. This function should be generic and not specific to DSR

ksubrmnn commented on pkg/proxy/winkernel/hnsV2.go:
```diff
@@ -89,18 +91,31 @@ func (hns hnsV2) getEndpointByIpAddress(ip string, networkName string) (*endpoin
 
 	return nil, fmt.Errorf("Endpoint %v not found on network %s", ip, networkName)
 }
+func transformToPoliciesinfo(hcnendpointpolicies []hcn.EndpointPolicy) []*policiesinfo {
```
Rename to policiesToPolicyInfo

ksubrmnn commented on pkg/proxy/winkernel/hnsV2.go:
```diff
@@ -89,18 +91,31 @@ func (hns hnsV2) getEndpointByIpAddress(ip string, networkName string) (*endpoin
+func transformToPoliciesinfo(hcnendpointpolicies []hcn.EndpointPolicy) []*policiesinfo {
+	var endpointPolicies []*policiesinfo
+	for _, po := range hcnendpointpolicies {
+		var policy policiesinfo
```
Include a log statement here for po

zawachte-msft replied:
There is one for policy on line 115 ... but not sure if you were suggesting that we should have a log for both structs?

ksubrmnn replied:
Yes we should log po (each policy we inspect) as well as the policyInfo we create

ksubrmnn commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -218,7 +237,7 @@ func newServiceInfo(svcPortName proxy.ServicePortName, port *v1.ServicePort, ser
 		loadBalancerSourceRanges: make([]string, len(service.Spec.LoadBalancerSourceRanges)),
 		onlyNodeLocalEndpoints:   onlyNodeLocalEndpoints,
-		hns:                      hns,
+		hns: hns,
```
Does this pass gofmt?

zawachte-msft replied:
gofmt made this change.

jeefy committed:
spelling/grammar fixes

ksubrmnn commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -1025,6 +1044,51 @@ func (proxier *Proxier) syncProxyRules() {
 
+			if newHnsEndpoint != nil {
+				if newHnsEndpoint.isLocal && proxier.isDSR {
+					var alreadyHasDSRPolicy = false
```
Rename to addDSRPolicy

ksubrmnn commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -1025,6 +1044,51 @@ func (proxier *Proxier) syncProxyRules() {
+						if strings.Compare(po.Type, "OutBoundNAT") == 0 {
+							var outputSettings outboundnatpolicysetting
+							if err := json.Unmarshal([]byte(po.Settings), &outputSettings); err != nil {
+								break
```
Won't this break out of the loop meaning you might not check the other policies?

ksubrmnn commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -24,6 +24,7 @@ import (
 	"net"
 	"os"
 	"reflect"
+	"strings"
```
You can just use == to compare strings in Go. Is there a reason you want to use strings?

zawachte-msft replied:
Using == makes me queasy. Especially if this is code that may be running cross-platform. But if that is the go standard... I can change it back.

ksubrmnn replied:
== is fine. There’s no sort of user input here

ksubrmnn commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -1025,6 +1044,51 @@ func (proxier *Proxier) syncProxyRules() {
 			}
 
+			if newHnsEndpoint != nil {
+				if newHnsEndpoint.isLocal && proxier.isDSR {
```
Include a comment for this block explaining that this fixes the DSR loopback issue.

ksubrmnn commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -1025,6 +1044,51 @@ func (proxier *Proxier) syncProxyRules() {
+
+						settingsJson, err := json.Marshal(endpointRequest)
+						if err != nil {
+							break
```
Log this error

ksubrmnn replied:
I think this break will skip the logging, ep Info updating in 1150 etc.

ksubrmnn commented:
/ok-to-test

smarterclayton committed:
Avoid allocations when building SelfLinks and fast path escape

A self link should only require one allocation, and we should skip
url.PathEscape() except when the path actually needs it.

Add a fuzz test to build random strings and verify them against
the optimized implementation. Add a new BenchmarkWatchHTTP_UTF8 that
covers when we have unicode names in the self link.

```
> before
BenchmarkGet-12          	   10000	    118863 ns/op	   17482 B/op	     130 allocs/op
BenchmarkWatchHTTP-12    	   30000	     38346 ns/op	    1893 B/op	      29 allocs/op

> after
BenchmarkGet-12               	   10000	    116218 ns/op	   17456 B/op	     130 allocs/op
BenchmarkWatchHTTP-12         	   50000	     35988 ns/op	    1571 B/op	      26 allocs/op
BenchmarkWatchHTTP_UTF8-12    	   50000	     41467 ns/op	    1928 B/op	      28 allocs/op
```

Saves 3 allocations in the fast path and 1 in the slow path (the
slow path has to build the buffer and then call url.EscapedPath
which always allocates).

Commit:
cleanup of reflector metric code (finish removing unused code)

k8s-ci-robot committed:
Merge pull request #75696 from jeefy/1.14-relnotes

Spelling/grammar fixes for 1.14 release notes

k8s-ci-robot committed:
Merge pull request #75623 from oomichi/golint-e2e-framework-k-l

Fix golint failures of e2e/framework/[k-l]*.go

k8s-ci-robot committed:
Merge pull request #75657 from guilhermeoki/docs/sample-controller

sample-controller: update README

Commit:
merging

k8s-ci-robot committed:
Merge pull request #75640 from fabriziopandini/e2e-kubeadm-refactor

e2e kubeadm refactor

k8s-ci-robot committed:
Merge pull request #75690 from oomichi/golint-e2e-framework-d-e

Fix golint failures of e2e/framework/[d-e]*.go

feiskyer commented:
@zawachte-msft This is cool. I think the changes don't cover https://github.com/kubernetes/kubernetes/pull/62093 yet, right? Could you like to cherry pick the commits together, or let me rebase #62093 after this one?

smarterclayton committed:
Make RequestScope be a pointer consistently for better memory use

RequestScope is a large struct and causes stack growth when we pass
it by value into multiple stack levels. Avoid the allocations for
this read only struct by passing a pointer.

k8s-ci-robot committed:
Merge pull request #75650 from xichengliudui/add-xichengliudui-to-onwers

Add @xichengliudui to test/OWNERS

k8s-ci-robot committed:
Merge pull request #75678 from javier-b-perez/debian-base

Bump debian-base images version to v1.0.0

danielqsj committed:
add completed job status in cronjob event

k8s-ci-robot committed:
Merge pull request #75685 from freehan/flaky-test

make NEG tests more resilient to API failures and add more failure lo…

k8s-ci-robot committed:
Merge pull request #75700 from logicalhan/reflector-cleanup

cleanup of reflector metric code (finish removing unused code)

MansM committed:
update changelog

update changelog

fabriziopandini committed:
e2e-kubeadm-new-test

wojtek-t committed:
Avoid allocations in ByIndex() function

ksubrmnn commented:
@feiskyer I think we should keep these PRs separate. Supporting only node local endpoints is different than fixing the loopback issue. 

Also looks like your PR will need rebasing anyway to use the HostNetworkService interface. Can we wait until Zach’s PR goes in and the you can rebase?

ksubrmnn commented:
@zawachte-msft 

Can you change the description for this PR and the release note?  

WinDSR as a feature flag is already functional and in alpha as of 1.14. The issue is that requests load balanced to the source container (loopback) fail. 

“Implementing DSR” is already handled by using proxies.isDSR when creating loadbalancers. This PR is more of a bug fix.

andrewsykim committed:
fix link to etcd install docs

k8s-ci-robot committed:
Merge pull request #75168 from ialidzhikov/update-golang-version

Update golang to 1.12.0

k8s-ci-robot committed:
Merge pull request #75693 from smarterclayton/speed_up_time_json

Reduce allocations in metav1.Time JSON serialization

Commit:
log out metric registration errors instead of silently ignoring them

Huang-Wei committed:
scheduler: correct dated comments on pod preemption

- correct wordings related with annotations - which hasn't been used since alpha version

k8s-ci-robot committed:
Merge pull request #72179 from WanLinghao/sa_controller

Migrate the controller to use TokenRequest and rotate token periodically

k8s-ci-robot committed:
Merge pull request #74341 from paulsubrata55/kube_proxy_sctp_fix

Issue in kube-proxy when IPVS is enabled and SCTP traffic is sent.

k8s-ci-robot committed:
Merge pull request #74616 from ialidzhikov/fluentd-1.4.0

Update fluentd to 1.4.0

k8s-ci-robot committed:
Merge pull request #75399 from sttts/sttts-initial-bulk-openapi-merge

apiextensions: merge OpenAPI specs once after cache sync

k8s-ci-robot committed:
Merge pull request #75515 from dougm/vsphere-token-auth

vSphere: add token auth support for tags client

Commit:
log preshutdowndown registration error if we encounter one

k8s-ci-robot committed:
Merge pull request #75445 from shinytang6/enhance/fmt

Replace all time.Now().Sub with time.Since

verult committed:
Updated regional PD minimum size; changed regional PD failover test to use StorageClassTest to generate PVC template

k8s-ci-robot committed:
Merge pull request #74678 from pmcalpine/lint-bs_cj

Fix some golint failures for pkg/controller

k8s-ci-robot committed:
Merge pull request #75699 from smarterclayton/fast_encode

Avoid allocations when building SelfLinks and fast path escape

feiskyer commented:
> Also looks like your PR will need rebasing anyway to use the HostNetworkService interface. Can we wait until Zach’s PR goes in and the you can rebase?

sure.

k8s-ci-robot committed:
Merge pull request #75737 from logicalhan/metric-errors

log out metric registration errors instead of silently ignoring them

k8s-ci-robot committed:
Merge pull request #75746 from verult/repd-min-size

Updated regional PD minimum size for zonal failover test

Commit:
PR feedback

BenTheElder committed:
fix kube::golang::is_instrumented_package

Commit:
rebase

k8s-ci-robot committed:
Merge pull request #75531 from dims/add-new-staging-repository-for-cri-api

New staging repository for cri-api

k8s-ci-robot committed:
Merge pull request #75717 from wojtek-t/reduce_critical_sections

Reduce critical sections in cacher::Watch function

k8s-ci-robot committed:
Merge pull request #75730 from wojtek-t/minor_perf_improvements

Avoid allocations in ByIndex() function

k8s-ci-robot committed:
Merge pull request #75735 from andrewsykim/update-install-etcd-link

fix link to etcd install docs for integration tests

k8s-ci-robot committed:
Merge pull request #75739 from Huang-Wei/cleanup-preemption-comments

scheduler: correct dated comments on pod preemption

k8s-ci-robot committed:
Merge pull request #75743 from logicalhan/preshutdownerror

log preshutdowndown registration error if we encounter one

denkensk committed:
clean up func podTimestamp

smarterclayton committed:
Avoid allocating the watch shim object more than once

We can reset and reuse this object repeatedly without triggering an
allocation.

k8s-ci-robot committed:
Merge pull request #74693 from mkimuram/issue/74545

Add e2e tests for multiAttach

k8s-ci-robot committed:
Merge pull request #75668 from bart0sh/PR0068-hack-lib-golang.sh-fix-mac-build

fix 'make generated_files' build on MacOS

k8s-ci-robot committed:
Merge pull request #75605 from smarterclayton/check_crd_proto

Check CRD protobuf negotiation in integration tests

k8s-ci-robot committed:
Merge pull request #75680 from tallclair/psp-refactor

Clean up some PodSecurityPolicy code

k8s-ci-robot committed:
Merge pull request #75718 from MansM/patch-1

update changelog

k8s-ci-robot committed:
Merge pull request #75751 from BenTheElder/bash-fun

fix kube::golang::is_instrumented_package

WanLinghao committed:
Fix a test build failed error

wojtek-t committed:
Add explanation about forgetFunc in cacher

k8s-ci-robot committed:
Merge pull request #75755 from smarterclayton/save_one_allocate

Avoid allocating the watch shim object more than once

tomwans committed:
Avoid panic in cronjob sorting

This change handles the case where the ith cronjob may have its start
time set to nil.

Previously, the Less method could cause a panic in case the ith
cronjob had its start time set to nil, but the jth cronjob did not. It
would panic when calling Before on a nil StartTime.

andrewsykim committed:
add @andrewsykim and @mcrute as pkg/credentialprovider revierws

Signed-off-by: Andrew Sy Kim <kiman@vmware.com>

k8s-ci-robot committed:
Merge pull request #74747 from liggitt/quota-deadlock

quota controller fixes

roycaihw committed:
kube-aggregator: bump openapi aggregation log level

k8s-ci-robot committed:
Merge pull request #72454 from rojkov/kubeadm-fallback-version

kubeadm: add final fallback to constants.CurrentKubernetesVersion

k8s-ci-robot committed:
Merge pull request #75773 from andrewsykim/credential-provider-owners

add @andrewsykim and @mcrute as pkg/credentialprovider reviewers

bart0sh committed:
kubeadm: use t.Run in app/phases

Used T.Run API for kubeadm tests in app/phases/*

This should improve testing output and make it more visible
which test is doing what.

Signed-off-by: Ed Bartosh <eduard.bartosh@intel.com>

k8s-ci-robot committed:
Merge pull request #75143 from rramkumar1/remove-http2-tests

Remove tests exercising the app protocol feature of Ingress-GCE

msau42 committed:
Switch PD tests to use delayed binding

k8s-ci-robot committed:
Merge pull request #72384 from zhouhaibing089/quota-crd

resourcequota: use dynamic informer

k8s-ci-robot committed:
Merge pull request #74399 from RA489/runpullimagescleanup

 Rename RunPullImagesCheck to PullControlPlaneImages

k8s-ci-robot committed:
Merge pull request #73380 from vllry/kube-proxy-cleanup

Superficial kube-proxy cleanup

k8s-ci-robot committed:
Merge pull request #74802 from WanLinghao/zh_bindata_fix

Remove 404 link in kubectl help information

msau42 committed:
return empty devicepath for csi attach

msau42 committed:
Fix provisioning tests to allow delayed binding

k8s-ci-robot committed:
Merge pull request #75510 from hex108/starttime

Pick pods for preemption based on StartTime of pods when priorities a…

k8s-ci-robot committed:
Merge pull request #75513 from cwdsuzhou/fix-incline-csivolume

Kubectl: support inline csi volume

k8s-ci-robot committed:
Merge pull request #75748 from smarterclayton/request_scope_as_pointer

Make RequestScope be a pointer consistently for better memory use

k8s-ci-robot committed:
Merge pull request #75760 from wojtek-t/follow_comment_from_critical_sections

Add explanation about forgetFunc in cacher

SataQiu committed:
fix none feature gates list

k8s-ci-robot committed:
Merge pull request #75784 from bart0sh/PR0069-kubeadm-use-t.Run-in-app-phases

kubeadm: use t.Run in app/phases

k8s-ci-robot committed:
Merge pull request #75641 from fabriziopandini/e2e-kubeadm-new-test

E2e kubeadm new test

Commit:
Adds DeviceMountablePlugin.CanDeviceMount check when retrieving plugins

k8s-ci-robot committed:
Merge pull request #75762 from SataQiu/fix-kubeadm-gates-20190327

Fix none feature gates list

k8s-ci-robot committed:
Merge pull request #73937 from smarterclayton/report_errors

Report a watch error instead of eating it when we can't decode

k8s-ci-robot committed:
Merge pull request #75362 from serathius/gcp-security-patches

Update gcp images with security patches

fabriziopandini committed:
fix-external-etcd

k8s-ci-robot committed:
Merge pull request #75847 from fabriziopandini/fix-external-etcd

kubeadm: fix join control-plane with external-etcd

pjh committed:
Disable GCE agent address management on Windows nodes.

With this metadata key set, "GCEWindowsAgent: GCE address manager
status: disabled" will appear in the VM's serial port output during
boot.

Tested:
PROJECT=${CLOUDSDK_CORE_PROJECT} KUBE_GCE_ENABLE_IP_ALIASES=true NUM_WINDOWS_NODES=2 NUM_NODES=2 KUBERNETES_NODE_PLATFORM=windows go run ./hack/e2e.go -- --up
cluster/gce/windows/smoke-test.sh

cat > iis.yaml <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: iis
  labels:
    app: iis
spec:
  containers:
  - image: mcr.microsoft.com/windows/servercore/iis
    imagePullPolicy: IfNotPresent
    name: iis-server
    ports:
    - containerPort: 80
      protocol: TCP
  nodeSelector:
    beta.kubernetes.io/os: windows
  tolerations:
  - effect: NoSchedule
    key: node.kubernetes.io/os
    operator: Equal
    value: windows1809
EOF

kubectl create -f iis.yaml
kubectl expose pod iis --type=LoadBalancer --name=iis
kubectl get services
curl http://<service external IP address>

Commit:
Fix golint failures of e2e/framework/n*.go

k8s-ci-robot committed:
Merge pull request #74844 from leakingtapan/migration-raw-block

Implement CSI migration logic for block volume

danielqsj committed:
support both JSON and YAML for scheduler configuration

shinytang6 committed:
scheduler: handle missing err

danielqsj committed:
add yaml test cases for scheduler creation

Commit:
Optimize descrive pvc output organization

SataQiu committed:
fix golint failures in cmd/kubeadm/app/util/system

xichengliudui committed:
Update OWNERS

k8s-ci-robot committed:
Merge pull request #75855 from pjh/gce-disable-agent-address-mgmt

Disable GCE agent address management on Windows nodes.

SataQiu committed:
fix golint failures of e2e/framework/test_context.go

SataQiu committed:
refactor code about setting KubernetesVersionFlag

hormes committed:
Refactor watchcache to support bookmark

k8s-ci-robot committed:
Merge pull request #75871 from SataQiu/fix-test-golint-20190329

Fix golint failures of e2e/framework/test_context.go

k8s-ci-robot committed:
Merge pull request #75799 from msau42/csi-empty-devicepath

return empty devicepath for csi attach

k8s-ci-robot committed:
Merge pull request #75879 from hormes/watchcache_refactor

Refactor watchcache to pass function to create objects instead of the empty object  itself

rramkumar1 committed:
Remove static IP test from Ingress-GCE e2e tests

k8s-ci-robot committed:
Merge pull request #74695 from qingsenLi/k8s190228-delete-code

delete unused code in proxy_server.go

k8s-ci-robot committed:
Merge pull request #75165 from qingsenLi/k8s190308-fix-error

fix a mistake in get.go

k8s-ci-robot committed:
Merge pull request #75584 from SataQiu/fix-golint-20190322

fix some golint failures in pkg/apis

Commit:
Fix golint failures of e2e/framework/p*.go

This fixes golint failures of
- test/e2e/framework/perf_util.go
- test/e2e/framework/pods.go
- test/e2e/framework/profile_gatherer.go
- test/e2e/framework/provider.go
- test/e2e/framework/psp_util.go

k8s-ci-robot committed:
Merge pull request #75210 from deads2k/scale-toleration

make kubectl scale work without a GET if a precodition isn't requested

k8s-ci-robot committed:
Merge pull request #75646 from bart0sh/PR0066-kubeadm-token-delete-multiple

kubeadm: implement deletion of multiple tokens

k8s-ci-robot committed:
Merge pull request #75712 from danielqsj/cj

Add completed job status in cronjob event

k8s-ci-robot committed:
Merge pull request #75834 from jlucktay/golint_pkg/auth/authorizer/abac

Golint pkg/auth/authorizer/abac

k8s-ci-robot committed:
Merge pull request #75863 from cwdsuzhou/optimize-describe-pvc-organization

Optimize describe pvc output organization

neolit123 committed:
kubeadm: fix "upgrade plan" not working without k8s version

If the k8s version argument passed to "upgrade plan" is missing
the logic should perform the following actions:
- fetch a "stable" version from the internet.
- if that fails, fallback to the local client version.

Currentely the logic fails because the cfg.KubernetesVersion is
defaulted to the version of the existing cluster, which
then causes an early exit without any ugprade suggestions.

See app/cmd/upgrade/common.go::enforceRequirements():
  configutil.FetchInitConfigurationFromCluster(..)

Fix that by passing the explicit user value that can also be "".
This will then make the "offline getter" treat it as an explicit
desired upgrade target.

In the future it might be best to invert this logic:
- if no user k8s version argument is passed - default to the kubeadm
version.
- if labels are passed (e.g. "stable"), fetch a version from the
internet.

k8s-ci-robot committed:
Merge pull request #75844 from vladimirvivien/volume-skip-device-mount

Volume DeviceMountablePlugin.CanDeviceMount check when retrieving plugins

ixdy committed:
bazel: update repo-infra dependency to fix bazel 0.24+ (requires 0.23+)

k8s-ci-robot committed:
Merge pull request #75587 from tiffanyfay/cred-provider

Refactor AWS credential provider

k8s-ci-robot committed:
Merge pull request #75627 from ialidzhikov/fix-lint-error

Fix lint issues

k8s-ci-robot committed:
Merge pull request #75481 from mucahitkurt/refactor/nodeinfomanager-tests

Add expectedCSINode for csiNodeEnabled test cases on nodeinfomanager tests

k8s-ci-robot committed:
Merge pull request #75754 from denkensk/cleanup-podtimestamp

clean up func podTimestamp in queue

k8s-ci-robot committed:
Merge pull request #75860 from oomichi/golint-e2e-framework-n

Fix golint failures of e2e/framework/n*.go

k8s-ci-robot committed:
Merge pull request #75861 from shinytang6/enhance/cleanup-scheduler

scheduler: handle err and remove useless assignment

k8s-ci-robot committed:
Merge pull request #75867 from SataQiu/fix-kubeadm-golint-20190329

Fix golint failures in cmd/kubeadm/app/util/system

k8s-ci-robot committed:
Merge pull request #75621 from oomichi/golint-e2e-framework-g-j

Fix golint failures of e2e/framework/[g-j]*.go

danielqsj committed:
add possible values to CustomResourceDefinitionConditionType

danielqsj committed:
update generated files

k8s-ci-robot committed:
Merge pull request #75613 from tallclair/owners

Copy node test owners to e2e/common

k8s-ci-robot committed:
Merge pull request #75868 from xichengliudui/patch-4

Add @xichengliudui to hack/OWNERS reviewers

k8s-ci-robot committed:
Merge pull request #75874 from SataQiu/refactor-kubeadm-20190329

Refactor code about setting KubernetesVersionFlag

k8s-ci-robot committed:
Merge pull request #75892 from rramkumar1/patch-14

Remove static IP test from Ingress-GCE e2e tests

k8s-ci-robot committed:
Merge pull request #75796 from msau42/fix-e2e-multizone

Switch PD tests to use delayed binding

hormes committed:
add timeout suuport for watch

k8s-ci-robot committed:
Merge pull request #75858 from danielqsj/crdct

add possible values to CustomResourceDefinitionConditionType

giuseppe committed:
test, e2e: add tests for logs -f

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

k8s-ci-robot committed:
Merge pull request #75900 from neolit123/fix-upgrade-plan

kubeadm: fix "upgrade plan" not defaulting to a "stable" version

k8s-ci-robot committed:
Merge pull request #75865 from xichengliudui/fixshellcheckfailure-03

Fix shellcheck failures in godep-save.sh and godep-restore.sh

k8s-ci-robot committed:
Merge pull request #75897 from oomichi/golint-e2e-framework-p

Fix golint failures of e2e/framework/p*.go

k8s-ci-robot committed:
Merge pull request #73672 from WanLinghao/failure_domain_remove

Remove dead code about failure-domains option in kube-scheduler

k8s-ci-robot committed:
Merge pull request #73737 from odino/owner-references-in-create-job

Attaching owner references to jobs create via "--from=cronjob/xyz"

k8s-ci-robot committed:
Merge pull request #75904 from ixdy/update-repo-infra-bazel-0.23

bazel: update repo-infra dependency to fix bazel 0.24+ (requires 0.23+)

SataQiu committed:
fix golint failures of e2e/framework/ssh.go, e2e/framework/size.go

k8s-ci-robot committed:
Merge pull request #75923 from SataQiu/fix-golint-ssh-size-20190331

Fix golint failures of e2e/framework/ssh.go, e2e/framework/size.go

danielqsj committed:
add unit test for initPolicyFromFile

k8s-ci-robot committed:
Merge pull request #75857 from danielqsj/sc1

support both JSON and YAML for scheduler configuration

k8s-ci-robot committed:
Merge pull request #75662 from s-ito-ts/shellcheck_local-up-cluster.sh

fix shellcheck failures in hack/local-up-cluster.sh

SataQiu committed:
use constant instead of hard coding

SataQiu committed:
fix golint failures of e2e/framework/pods.go, provider.go, upgrade_util.go

k8s-ci-robot committed:
Merge pull request #75595 from liggitt/client-go-instructions

Add preliminary go module instructions to client-go

k8s-ci-robot committed:
Merge pull request #75483 from gsadhani/describe-output-fix

Skip smart label in kubectl describe if field has special chars

k8s-ci-robot committed:
Merge pull request #75933 from SataQiu/fix-kubeadm-20190401

Update kubeadm/kubeconfig to use constant instead of hard coding

k8s-ci-robot committed:
Merge pull request #75835 from hormes/add_timeout_to_watch

add timeout support  for Watch()

k8s-ci-robot committed:
Merge pull request #75940 from SataQiu/fix-e2e-framework-20190401

Fix golint failures of e2e/framework/pods.go, provider.go, upgrade_util.go

k8s-ci-robot committed:
Merge pull request #75950 from qingsenLi/k8s190401-fix-Certifcate

spelling error 'Certifcate'

k8s-ci-robot committed:
Merge pull request #73041 from giuseppe/fix-logs

kubelet: fix some race conditions with logs -f

k8s-ci-robot committed:
Merge pull request #75759 from WanLinghao/api_meta_test_fix

Fix a test build failed error

k8s-ci-robot committed:
Merge pull request #75772 from tomwans/cronjob-sorting

Avoid panic in cronjob sorting

k8s-ci-robot committed:
Merge pull request #75781 from roycaihw/fix/openapi-aggregator-log

kube-aggregator: bump openapi aggregation log level for delegation targets

k8s-ci-robot committed:
Merge pull request #75939 from xichengliudui/conversion-constants

Define common sentences as constants

Commit:
spelling error 'Certifcate'

Commit:
switching branches

Commit:
merge
